### ISSUE TITLE
Subclass of native classes (re-consider or update docs)

### ISSUE THREAD

**Issue by tex0l** [electron/electron#8898]
I've just discovered that `BrowserWindow` and other native classes of Electron shouldn't be sub-classed, @zcbenz has provided an explanation for this choice in 2013 here: https://github.com/electron/electron/issues/23.

I am wondering if this can be reconsidered since it has been decided when ES2015 wasn't around, and that V8 has changed a lot since (I don't know if it's now possible). Today, it's very natural to write:
```javascript
class MyBrowserWindow extends BrowserWindow {
  constructor (options){
    super(options)
    // Do stuff
  }
}
const myBrowserWindowInstance = new MyBrowserWindow({ height:600; width: 800 })
```

In my case, I am currently sub-classing `BrowserWindow` with no found caveats since Electron 1.4.x, except when using methods that check the equality of `myBrowserWindow.constructor` with `BrowserWindow` (for example in `Menu` and `dialog` methods, see here https://github.com/electron/electron/issues/8895).

**If this is definitely a won't fix, the documentation should specify that native classes of Electron shouldn't be sub-classed.**

**Comment by joerick** [comment]
Agreed. I also assumed that subclassing would 'just work', being unaware of the complexities (as discussed in #23). Currently the failure mode is non-obvious (my app silently froze after showing a Save dialog). In addition to the docs update, perhaps we could add runtime checks to BrowserWindow to ensure it wasn't subclassed? e.g. 

    if(this.constructor !== BrowserWindow) {
        throw new Error('BrowserWindow cannot be subclassed')
    }

**Comment by cedx** [comment]
Encountered this issue today. This should be clearly mentioned somewhere in the docs.

**Comment by RoeelCohen** [comment]
Same here, in my case:

- BrowserWindows was created.
- Get browser window from ID using `BrowserWindow.fromId(id)` would always return a null.
- `BrowserWindow.getAllWindows()` always returns an empty Array.

I really thought I was losing it since I could see the windows with my own eyes.
Spent a lot of time and effort on this one

**Comment by RoeelCohen** [comment]
I extended `BrowserWindow` class to create popup and main window classes.

Although BrowserWindows was created successfully electron BrowserWindow didn't keep track of those windows.
That means that

- Get browser window from ID using `BrowserWindow.fromId(id)` would always return a null.
- `BrowserWindow.getAllWindows()` always returns an empty Array.

I couldn't find any issues with my code nor any word about it in the docs.

**Comment by theogravity** [comment]
Same, I thought I could get access to `BaseWindow` since `BrowserWindow` has that constructor check to re-implement `getFocusedWindow()` and `getAllWindows()` but that doesn't seem possible

**Comment by miss-programgamer** [comment]
Eight god-damned fucking years.

**Comment by miss-programgamer** [comment]
Oh I'm sorry, let me rephrase that then. Golly gee, eight whole years have passed since this was submitted! If someone were to trip over this pitfall and find that their troubles could have been avoided if someone did their job in a timely fashion, hoo boy would that cause some mighty frustration!

Fix your docs. Bitch.

**Comment by miss-programgamer** [comment]
One line at the top of one documentation page. Either go do it now or give me access to the documentation so I can do it myself.

Here, I'll even link it for you, just to be extra nice about it: [https://www.electronjs.org/docs/latest/api/browser-window](https://www.electronjs.org/docs/latest/api/browser-window)

**Comment by miss-programgamer** [comment]
I still don't see a line on that page about not subclassing `BrowserWindow`.

**Comment by Jackietkfrost** [comment]
Is there any intent on updating this documentation? I think it would've saved me a lot of time on some of the stuff I've been working on instead of stumbling on this random issue.
